stacktrace.go

```diff
@@ -20,23 +20,75 @@
 
 package zap
 
-import "runtime"
+import (
+	"runtime"
+	"strings"
+	"sync"
+
+	"go.uber.org/zap/internal/bufferpool"
+)
+
+var (
+	_stacktraceIgnorePrefixes = []string{
+		"go.uber.org/zap",
+		"runtime.",
+	}
+	_stacktraceUintptrPool = sync.Pool{
+		New: func() interface{} {
+			return make([]uintptr, 64)
+		},
+	}
+)
 
 // takeStacktrace attempts to use the provided byte slice to take a stacktrace.
 // If the provided slice isn't large enough, takeStacktrace will allocate
 // successively larger slices until it can capture the whole stack.
-func takeStacktrace(buf []byte, includeAllGoroutines bool) string {
-	if len(buf) == 0 {
-		// We may have been passed a nil byte slice.
-		buf = make([]byte, 1024)
+func takeStacktrace() string {
```

Comment:
Yep. In the context of taking a stacktrace, which is more than a microsecond, there's no harm in a few cheap string comparisons. I'd very, very much like to avoid another hard-coded caller skip, since it limits our options for structuring code in zap's internals - either every user-facing API must be the same number of stack frames away from this code, or we need to plumb skip arguments everywhere.